Implement STM32H73B3I_DK board and STM32H7A3 family MCUs

JohnieBraaf opened the pull request:
The groundwork for this PR was done by @dpgeorge [stm32-h7b3-support](https://github.com/dpgeorge/micropython/tree/stm32-h7b3-support) and @koendv [micropython-spiram](https://github.com/koendv/micropython-spiram)

I have integrated their work and added some fixes here and there.

**MCU A3/B3 support**
This PR adds basic support for the STM32H7A3 and B3 families MCU's:
- STM32H7A3xx
- STM32H7A3xxQ (SMPS)
- STM32H7B3xx
- STM32H7B3xxQ (SMPS)

For alternate functions stm32h7b3_af.csv has been added which provides mapping for the 2 available OCTO SPI interfaces. Micropython does not currently have a provision for this bus type however.

**STM32H7B3I_DK support**
Support has been added for the STM32H7B3I_DK board.

Currently functioning:
- REPL (via ST-Link usb)
- UART4 (tested)
- SDCard (tested)
- timers (tested)
- adc (tested)
- leds (tested)
- button (tested)

This PR has been [squashed](https://github.com/JohnieBraaf/micropython/commits/c4f4cb14a455c7dfbe81d18265b9cfb718d265f0) and tested by me and can be reviewed/altered/merged as is.

Some useful resources:
- [Reference manual STM32H7A3/B3 MCU](https://www.st.com/resource/en/reference_manual/dm00463927-stm32h7a37b3-and-stm32h7b0-value-line-advanced-armbased-32bit-mcus-stmicroelectronics.pdf)
- [User Manual STM32H7B3I_DK](https://www.st.com/resource/en/user_manual/dm00610478-discovery-kit-with-stm32h7b3li-mcu-stmicroelectronics.pdf)
- [Octo SPI application Note](https://www.st.com/resource/en/application_note/dm00407776-octospi-interface-on-stm32-microcontrollers-stmicroelectronics.pdf)
- Sample project by ST - NOR Flash using OCTO SPI: [MX25LM51245G_STM32H7B3I-DISCO](https://github.com/STMicroelectronics/stm32-external-loader/tree/main/STM32H7x_boards/MX25LM51245G_STM32H7B3I-DISCO)

JohnieBraaf commented:
Note to @dpgeorge , SMPS MCU's can be deadlocked, making a board unflashable when wrong clock configuration is given.
There is a special restore procedure involving the swith left of the screen (page 18 of [UM2569](https://www.st.com/resource/en/user_manual/dm00610478-discovery-kit-with-stm32h7b3li-mcu-stmicroelectronics.pdf)) in order to recover the board 

```
If a deadlock is faced due to a mismatch between the hardware board setting and the firmware
setting (LDO/SMPS), the user can recover the board by doing the following:
• Power off the board.
• Change the SW1 switch position to "1".
• This changes the BOOT0 pin to 1 instead of 0 and thus the device boot address is changed
to boot address 1 making the bootloader starting in System memory, instead of starting the
firmware in the user Flash (Firmware that is setting a wrong LDO/SMPS configuration).
• Power on the board, and connect using STM32CubeProgrammer.
• Erase the user Flash.
• Power off the board and put back the switch SW1 to position "0".
• The board is now recovered and can proceed normally. 
```

dpgeorge commented:
Thanks @JohnieBraaf for the info about deadlocked boards.  That did help to get STM32CubeProgrammer working again (able to erase and flash the board), but MicroPython still does not give a REPL.  Not sure why.

Anyway, I tried this PR on a new DK board and it works well.

JohnieBraaf commented:
Now uncrustified and in line with commit guidelines.

JohnieBraaf commented:
Sorry, first time contributurer to Micropython and unfortunately had an issue in my regex commit message checker.
Now, updated with capitalized first characted after the :

Let's just keep it at third time's the charm. ^_^

dpgeorge commented:
> Sorry, first time contributurer to Micropython and unfortunately had an issue in my regex commit message checker.

No worries, the CI can be very strict!

dpgeorge commented:
> and @koendv [micropython-spiram](https://github.com/koendv/micropython-spiram)

I can't see any spiram/octospi support in this PR, so what exactly did you take from that repo?  (The main reason to ask is because of licensing, that repo has no license).

dpgeorge commented on ports/stm32/boards/STM32H7B3I_DK/mpconfigboard.h:
```diff
@@ -0,0 +1,116 @@
+#endif
+
+// SPI buses
+#if 0
```
as with I2C, best to remove this if it's unused

dpgeorge commented on ports/stm32/boards/STM32H7B3I_DK/mpconfigboard.h:
```diff
@@ -0,0 +1,116 @@
+
+// USB config
+//#define MICROPY_HW_USB_HS_ULPI_NXT  (pin_C3)
+//#define MICROPY_HW_USB_HS_ULPI_DIR  (pin_C2)
```
these two lines can be removed

JohnieBraaf commented:
The code that I've integrated is inside the patch https://github.com/koendv/micropython-spiram/blob/main/stm32h7a3.patch

I have asked permission from the author by email.

```
Hi Jan,

No problem. I'm happy to see my code is put to good use.

best,

Koen
```

dpgeorge commented on ports/stm32/system_stm32.c:
```diff
@@ -195,8 +197,11 @@ void SystemClock_Config(void) {
-    #if defined(STM32H7)
     // Wait for PWR_FLAG_VOSRDY
+    #if defined(STM32H7A3xx) || defined(STM32H7A3xxQ) || defined(STM32H7B3xx) || defined(STM32H7B3xxQ)
+    while (!__HAL_PWR_GET_FLAG(PWR_FLAG_VOSRDY)) {
```
looks like this macro will work for all H7's, so could simplify this code block to:
```c
#if defined(STM32H7)
while (!__HAL_PWR_GET_FLAG(PWR_FLAG_VOSRDY)) {
}
#endif
```

dpgeorge commented on ports/stm32/usbd_conf.c:
```diff
@@ -127,10 +128,16 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd) {
+    else
+    #endif
+    #endif
+    #if MICROPY_HW_USB_HS
```
instead of this guarded `else`, would be simpler to just put a `return;` at the end of the `USB_OTG_FS` if block

dpgeorge commented on ports/stm32/usbd_conf.c:
```diff
@@ -203,14 +210,18 @@ void HAL_PCD_MspDeInit(PCD_HandleTypeDef *hpcd) {
-    else if (hpcd->Instance == USB_OTG_HS) {
+    else
+    #endif
+    #endif
```
similar to above, can just put a `return;` in the USB_OTG_FS if-block

JohnieBraaf commented:
B.t.w. indeed the author seems to use the octo-spi interface in sdr mode (qspi).

```
hospi1.Init.DualQuad = HAL_OSPI_DUALQUAD_DISABLE;
hospi1.Init.MemoryType = HAL_OSPI_MEMTYPE_APMEMORY; // sdr qspi
```

dpgeorge commented:
> The code that I've integrated is inside the patch

Aah, I see.  Ok, that's good then.

JohnieBraaf commented on ports/stm32/Makefile:
```diff
@@ -77,7 +77,7 @@ CFLAGS_CORTEX_M = -mthumb
 SUPPORTS_HARDWARE_FP_SINGLE = 0
 SUPPORTS_HARDWARE_FP_DOUBLE = 0
-ifeq ($(CMSIS_MCU),$(filter $(CMSIS_MCU),STM32F765xx STM32F767xx STM32F769xx STM32H743xx STM32H750xx))
+ifeq ($(CMSIS_MCU),$(filter $(CMSIS_MCU),STM32F765xx STM32F767xx STM32F769xx STM32H743xx STM32H750xx STM32H7A3xx STM32H7A3xxQ STM32H7B3xx STM32H7B3xxQ))
```
Added this part in the latest commit

dpgeorge replied:
Looks good!

JohnieBraaf commented:
Hi @dpgeorge, thank you for your review. I agree with your remarks and have updated the PR accordingly.

JohnieBraaf committed:
stm32/STM32H73B3I_DK: add board and support STM32H7A3 B3 MCUs.

This commit is based upon prior work of dpgeorge and koendv.

MCU support for the STM32H7A3 and B3 families MCUs:
- STM32H7A3xx
- STM32H7A3xxQ (SMPS)
- STM32H7B3xx
- STM32H7B3xxQ (SMPS)

Support has been added for the STM32H7B3I_DK board.

Signed-off-by: Jan Staal <info@janstaal.com>

JohnieBraaf commented:
Rebased

JohnieBraaf commented on ports/stm32/powerctrl.c:
```diff
@@ -36,14 +36,22 @@
 #elif defined(STM32H747xx)
 #define RCC_SR_SFTRSTF  RCC_RSR_SFT2RSTF
+#elif defined(STM32H7B3xx) || defined(STM32H7B3xxQ)
+#define RCC_SR_SFTRSTF  RCC_RSR_SFTRSTF // TODO check
```
Checked

JohnieBraaf commented on ports/stm32/boards/STM32H7B3I_DK/mpconfigboard.h:
```diff
@@ -0,0 +1,116 @@
+
+#if 0
+// 512MBit external OSPI flash, used for either the filesystem or XIP memory mapped
+#define MICROPY_HW_OSPIFLASH_SIZE_BITS_LOG2 (29) // ???
```
was a personal note, now removed

dpgeorge commented:
Thanks for updating.  Merged in 9e2423e730ae32cd82e5c06099e6a847a2020f6a with some minor changes:
- removed 1st column from pins.csv for pins that did not have a board name
- fixed location of return statements in usbd_conf.c

dpgeorge commented on ports/stm32/usbd_conf.c:
```diff
@@ -126,11 +127,16 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd) {
         HAL_NVIC_EnableIRQ(OTG_FS_IRQn);
         #endif
     }
+    return;
```
the return needed to go inside the if (fixed during merge)

JohnieBraaf commented:
> * removed 1st column from pins.csv for pins that did not have a board name

Thank you, ok, will keep that in mind

> * fixed location of return statements in usbd_conf.c

Yes, absolutelely (in hindsight how can I have missed that)

Happy to see this merged in any case.
